app/models/setting.rb

```diff
@@ -85,76 +68,104 @@ class Setting < ApplicationRecord
+          ActiveRecord::Type::Boolean.new.cast(value)
+        end
+
+        def self.#{name}=(value)
+          if settings_table_exists_yet?
+            self[:#{name}] = value
+          else
+            logger.warn "Trying to save a setting named '#{name}' while there is no 'setting' table yet. This setting will not be saved!"
+            nil # when running too early, there is no settings table. do nothing
+          end
+        end
 
-      def self.#{name}=(value)
-        if settings_table_exists_yet?
-          self[:#{name}] = value
-        else
-          logger.warn "Trying to save a setting named '#{name}' while there is no 'setting' table yet. This setting will not be saved!"
-          nil # when running too early, there is no settings table. do nothing
+        def self.#{name}_writable?
+          Settings::Definition[:#{name}].writable?
         end
+      END_SRC
+      class_eval src, __FILE__, __LINE__
+    end
+
+    def definitions
+      Settings::Definition.all
+    end
+
+    def method_missing(method, *args, &block)
+      if exists?(accessor_base_name(method))
+        create_setting_accessors(accessor_base_name(method))
+
+        send(method, *args)
+      else
+        super
       end
-    END_SRC
-    class_eval src, __FILE__, __LINE__
-  end
+    end
 
-  @@available_settings = YAML::load(File.open(Rails.root.join('config/settings.yml')))
+    def respond_to_missing?(method_name, include_private = false)
+      exists?(accessor_base_name(method_name)) || super
+    end
+
+    private
 
-  # Defines getter and setter for each setting
-  # Then setting values can be read using: Setting.some_setting_name
-  # or set using Setting.some_setting_name = "some value"
-  @@available_settings.each do |name, _params|
-    create_setting_accessors(name)
+    def accessor_base_name(name)
+      name.to_s.sub(/(_writable\?)|(\?)|=\z/, '')
+    end
   end
 
   validates_uniqueness_of :name
-  validates_inclusion_of :name, in: lambda { |_setting|
-                                      @@available_settings.keys
-                                    } # lambda, because @available_settings changes at runtime
-  validates_numericality_of :value, only_integer: true, if: Proc.new { |setting|
-                                                              @@available_settings[setting.name]['format'] == 'int'
-                                                            }
+  validates :name,
+            inclusion: {
+              in: ->(*) { Settings::Definition.all.map(&:name) } # @available_settings change at runtime
+            }
+  validates :value,
+            numericality: {
+              only_integer: true,
+              if: Proc.new { |setting| setting.format == :integer }
+            }
 
   def value
     self.class.deserialize(name, read_attribute(:value))
   end
 
-  def value=(v)
-    write_attribute(:value, formatted_value(v))
+  def value=(val)
+    raise NoMethodError unless Settings::Definition[name].writable?
```

Comment:
May I suggest this for a better DX:
```suggestion
    raise NoMethodError, "#{name} is not writable" unless Settings::Definition[name].writable?
```

I also thought of adding `", it can only be defined through env vars or config file"`. It feels helpful and too long at the same time.

Reply:
Done in [056fce4](https://github.com/opf/openproject/pull/10296/commits/056fce4b1921d637dcb00fa9c1314c1cd9c668f8). Thanks for the suggestion.

Reply:
Perfect! Thanks